core/unit_test/TestViewMapping_a.hpp

```diff
@@ -1066,10 +1066,7 @@ void test_view_mapping() {
     ASSERT_EQ(a.use_count(), 1);
     ASSERT_EQ(b.use_count(), 0);
 
-// TODO: a.use_count() and x.use_count() are 0 with the asynchronous HPX
-// backend. Why?
-#if !defined(KOKKOS_ENABLE_CUDA_LAMBDA) && \
-    !(defined(KOKKOS_ENABLE_HPX) && defined(KOKKOS_ENABLE_HPX_ASYNC_DISPATCH))
+#if !defined(KOKKOS_ENABLE_CUDA_LAMBDA)
```

Comment:
See https://github.com/kokkos/kokkos/pull/5628/commits/81b07df45ace27fdcbb1d04a50f5caab8cad3b13.

Reply:
Hmm, should it be `#if !defined(KOKKOS_ENABLE_CXX11_DISPATCH_LAMBDA)`? Or some combination with `*_CUDA` as well? The test is failing right now with CUDA (e.g. https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/11703/pipeline/54#step-329-log-598). I don't really understand what the `ifdef` is supposed to guard against...

Reply:
The macro is guarding against `CUDA` not supporting `__host__ __device__` lambdas without an extra compiler flag. `Cuda` is the only backend that doesn't support that natively.
If you need more restrictions for `HPX` here, you should add them.

Reply:
I don't need the condition for HPX anymore. However, CUDA clearly needs it. I see `KOKKOS_ENABLE_CXX11_DISPATCH_LAMBDA` is unconditionally defined here: https://github.com/kokkos/kokkos/blob/5ad609661e570ba6aa7716a26a91cb67d559f8a2/core/src/Kokkos_Macros.hpp#L108.

Based on the comment I _think_ it should say `#if !defined(KOKKOS_ENABLE_CUDA) || !defined(KOKKOS_ENABLE_CUDA_LAMBDA)` (or equivalently `#if !(defined(KOKKOS_ENABLE_CUDA) && defined(KOKKOS_ENABLE_CUDA_LAMBDA))`). I've updated it to the first, let's see if CI agrees with that.

Reply:
The configuration for the `CUDA` backend can be found at https://github.com/kokkos/kokkos/blob/5ad609661e570ba6aa7716a26a91cb67d559f8a2/core/src/setup/Kokkos_Setup_Cuda.hpp#L84-L93.
In any case, it would be a compile-time error if the guard isn't set correctly not a run-time error.

Reply:
I don't think the ifdef was there to guard against it not compiling. There's clearly a semantic difference between CUDA's experimental lambdas enabled os disabled, but I don't know where it comes from. The original ifdef goes back to https://github.com/kokkos/kokkos/pull/197 (which doesn't have any explanation about why it was disabled). Note that the condition is to enable the test when CUDA lambdas are _disabled_. FWIW the test is now disabled and the CUDA build is green again.

Reply:
Ah, yes, I misread the guard to mean what we normally check. I think the current version is OK then and any change would be out-of-scope for this pull request.